### ISSUE TITLE
bdev nvme reconnection does not work as expected when connecting to remote NVMe-oF target

### ISSUE THREAD

**Issue by optimistyzy** [spdk/spdk#668]
Please use the issue tracker only for reporting suspected issues.

See [The SPDK Community Page](http://www.spdk.io/community/) for other SPDK communications channels.

There are following two cases, we are not handling in a correct way when using SPDK NVMe bdev (as initiator) to connect to the remote NVMe-oF target.

Case 1  If we shut down remote the NVMe-oF target,  and restart the NVMe-oF target, the SPDK application which leverages Bdev NVMe and connect to the target can not resume the connection.  Since nvme_ctrlr_process_init is mainly considered for local NVMe devices. For RDMA or TCP transport, if the connection is broken, we should connect the NVMe-oF target again. However in our current code, we do not have such support.  There is no code in the reset path to reconnect to the remote NVMe-oF target.
Case 2 If we shut down the remote NVMe-oF target, and not start it again. The bdev NVMe should abort those I/Os due to time out. 

## Expected Behavior
Still consider the following two cases.

Case 1: If we shut  down the remote NVMe-of target, and restart it again.  The SPDK application which leverages NVMe bdev module could reconnect the NVMe-oF target, and resumes the I/O.

Case2: If we shut down the remote NVMe-oF target and not restart.  The SPDK application with NVMe bdev should abort those I/Os due to timeout.

## Current Behavior
Both Case 1 can Case2 are not satisfied.  I checked the code, currently the code path for spdk_nvme_ctrlr_reset is mainly considering for the local NVMe device, but not work for remote NVMe-of target.  We should considering adding some functions to enable auto I/O connection when the remote NVMe-of Target fails, some routines may be simliart to spdk_nvme_ctrlr_reset logic.

## Possible Solution
It seems a feature requirement. 

## Steps to Reproduce

1.
2.
3.
4.

**Comment by benlwalker** [comment]
Two separate improvements:

1. Implement automatic reconnect attempts when queue pairs drop their connection.
2. In the bdev nvme module, on an I/O timeout, need to go through a process like an spdk_bdev_reset command instead of just directly calling spdk_nvme_reset.

**Comment by optimistyzy** [comment]
@benlwalker  @jimharris  @Seth5141  

For 1:
I posted a patch to reconnect the qpair for remote transport and try to solve the reconnection issue:
https://review.gerrithub.io/#/c/spdk/spdk/+/445061/   And today @Seth5141  also provide a me a link which he abandoned, he also observed this issue before. 

For 2 the timeout function currently implemented in bdev nvme is not correct;
(1)About the qpair. 
If qpair == NULL, it is timeout caused by the admin qpair, then we should do the reset (I think that we should only let the admin qpair do the reset)
If the qpair != NULL, it is I/O qpair. So it seems that even if we want to do the reset, we should not do the brute force way. 

(2) if the controller is getting from the remote ctrlr, checking csts here is not exactly correct. Since for remote ctrlr we do not supporting get csts status (I checked the code for NVMe-oF target, get such property is not allowed). So better to differentiate the ctrlr with local or remote.

(3)Directly calling spdk_nvme_ctrlr_reset in this function is wrong. Supposing you are running with multiple threads (cores), we should do the similar things like bdev_nvme_reset function. Otherwise, you will trigger the problems (e.g., contention issue). Since in spdk_nvme_ctrlr_reset function, it will operate on all the I/O qpairs. However we know those I/O qpairs are created by different threads, so it must be handled by its created thread. So this is why it should be handled like a similar routine like bdev_nvme_reset
